browser/ui/views/infobars/web_discovery_infobar_view.cc

```diff
@@ -0,0 +1,51 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/views/infobars/web_discovery_infobar_view.h"
+
+#include <utility>
+
+#include "brave/browser/ui/views/infobars/web_discovery_infobar_content_view.h"
+#include "brave/browser/web_discovery/web_discovery_infobar_delegate.h"
+#include "chrome/browser/ui/views/frame/browser_view.h"
+#include "ui/base/metadata/metadata_impl_macros.h"
+#include "ui/views/widget/widget.h"
+
+std::unique_ptr<infobars::InfoBar> CreateWebDiscoveryInfoBar(
+    std::unique_ptr<WebDiscoveryInfoBarDelegate> delegate) {
+  return std::make_unique<WebDiscoveryInfoBarView>(std::move(delegate));
+}
+
+WebDiscoveryInfoBarView::WebDiscoveryInfoBarView(
+    std::unique_ptr<WebDiscoveryInfoBarDelegate> delegate)
+    : InfoBarView(std::move(delegate)) {
+  content_view_ = AddChildView(
+      std::make_unique<WebDiscoveryInfoBarContentView>(GetDelegate()));
+}
+
+WebDiscoveryInfoBarView::~WebDiscoveryInfoBarView() = default;
+
+WebDiscoveryInfoBarDelegate* WebDiscoveryInfoBarView::GetDelegate() {
+  return static_cast<WebDiscoveryInfoBarDelegate*>(delegate());
+}
+
+void WebDiscoveryInfoBarView::Layout() {
+  // Don't need to layout base class's elements as our |content_view_| covers
+  // all area except bottom separator.
+  content_view_->SetBounds(0, OffsetY(content_view_), width(),
+                           content_view_->height());
+}
+
+void WebDiscoveryInfoBarView::ChildPreferredSizeChanged(views::View* child) {
+  if (child == content_view_) {
+    child->SizeToPreferredSize();
+    RecalculateHeight();
+    BrowserView::GetBrowserViewForNativeWindow(GetWidget()->GetNativeWindow())
+        ->InvalidateLayout();
```

Comment:
Q. I'm wondering why we get browser view here. Doesn't WebDiscoveryInfoBarView::InvalidateLayout() work here?

Reply:
Ah, you're right. Actually both are same :) Updated.